pkg/cloudclient/gcp/private.go

```diff
@@ -0,0 +1,376 @@
+		c.logger.Debug(ctx, "Waiting on VM operation: %s", status)
+
+	case "STOPPING", "STOPPED", "TERMINATED", "SUSPENDED":
+		c.logger.Debug(ctx, "Fatal - Instance status: ", instanceName)
+		return "FATAL", fmt.Errorf(status)
+	}
+
+	if len(status) == 0 {
+		c.logger.Debug(ctx, "Instance %s has no status yet", instanceName)
+	}
+	return status, nil
+}
+
+func (c *Client) waitForComputeServiceInstanceCompletion(ctx context.Context, instanceName string) error {
+	//wait for the instance to run
+	err := helpers.PollImmediate(5*time.Second, 2*time.Minute, func() (bool, error) {
+		code, descError := c.describeComputeServiceInstances(ctx, instanceName)
+		switch code {
+		case "RUNNING":
+			//instance is running, break
+			c.logger.Info(ctx, "ComputeService Instance: %s %s", instanceName, code)
+			return true, nil
+
+		case "FATAL":
+			return false, fmt.Errorf("Instance %s already exists with %v state. Please run again", instanceName, descError)
+
+		case "PERMISSION DENIED":
+			return false, fmt.Errorf("missing required permissions for account: %v", descError)
+		}
+
+		if descError != nil {
+			return false, descError // unhandled
+		}
+
+		return false, nil // continue loop
+	})
+
+	return err
+}
+
+func generateUserData(variables map[string]string) (string, error) {
+	variableMapper := func(varName string) string {
+		return variables[varName]
+	}
+	data := os.Expand(helpers.UserdataTemplateGcp, variableMapper)
+
+	return data, nil
+}
+
+func (c *Client) findUnreachableEndpoints(ctx context.Context, instanceName string) error {
+	// Compile the regular expressions once
+	reVerify := regexp.MustCompile(userdataEndVerifier)
+	reUnreachableErrors := regexp.MustCompile(`Unable to reach (\S+)`)
+
+	// getConsoleOutput then parse, use c.output to store result of the execution
+	err := helpers.PollImmediate(30*time.Second, 4*time.Minute, func() (bool, error) {
+		output, err := c.computeService.Instances.GetSerialPortOutput(c.projectID, c.zone, instanceName).Context(ctx).Do()
+		if err != nil {
+			return false, err
+		}
+
+		if output != nil {
+			// First, gather the ComputeService console output
+			scriptOutput := fmt.Sprintf("%#v", output)
+
+			if err != nil {
+				// unable to decode output. we will try again
+				c.logger.Debug(ctx, "Error while collecting console output, will retry on next check interval: %s", err)
+				return false, nil
+			}
+
+			// In the early stages, an ComputeService instance may be running but the console is not populated with any data, retry if that is the case
+			if len(scriptOutput) < 1 {
+				c.logger.Debug(ctx, "ComputeService console output not yet populated with data, continuing to wait...")
+				return false, nil
+			}
+
+			// Check for the specific string we output in the generated userdata file at the end to verify the userdata script has run
+			// It is possible we get EC2 console output, but the userdata script has not yet completed.
+			verifyMatch := reVerify.FindString(string(scriptOutput))
```

Comment:
TODO: (please open a card) once we get the console output, the parsing and error checking can be common with aws/private.go. This logic could be moved to a helper

Reply:
Created https://issues.redhat.com/browse/OSD-12771